x-pack/plugins/discover_enhanced/public/actions/view_in_discover/explore_data_context_menu_action.ts

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/* eslint-disable max-classes-per-file */
+
+import { i18n } from '@kbn/i18n';
+import { Action } from '../../../../../../src/plugins/ui_actions/public';
+import { DiscoverStart } from '../../../../../../src/plugins/discover/public';
+import {
+  EmbeddableContext,
+  IEmbeddable,
+  ViewMode,
+} from '../../../../../../src/plugins/embeddable/public';
+import { StartServicesGetter } from '../../../../../../src/plugins/kibana_utils/public';
+import { CoreStart } from '../../../../../../src/core/public';
+import {
+  VisualizeEmbeddableContract,
+  VISUALIZE_EMBEDDABLE_TYPE,
+} from '../../../../../../src/plugins/visualizations/public';
+
+// TODO: Replace this logic with KibanaURL once it is available.
+// https://github.com/elastic/kibana/issues/64497
+class KibanaURL {
+  public readonly path: string;
+  public readonly appName: string;
+  public readonly appPath: string;
+
+  constructor(path: string) {
+    const match = path.match(/^.*\/app\/([^\/\/#]+)(.+)$/);
+
+    if (!match) {
+      throw new Error('Unexpected Discover URL path.');
+    }
+
+    const [, appName, appPath] = match;
+
+    if (!appName || !appPath) {
+      throw new Error('Could not parse Discover URL path.');
+    }
+
+    this.path = path;
+    this.appName = appName;
+    this.appPath = appPath;
+  }
+}
+
+export const ACTION_EXPLORE_DATA = 'ACTION_EXPLORE_DATA';
+
+const isOutputWithIndexPatterns = (
+  output: unknown
+): output is { indexPatterns: Array<{ id: string }> } => {
+  if (!output || typeof output !== 'object') return false;
+  return Array.isArray((output as any).indexPatterns);
+};
+
+const isVisualizeEmbeddable = (
+  embeddable: IEmbeddable
+): embeddable is VisualizeEmbeddableContract => embeddable?.type === VISUALIZE_EMBEDDABLE_TYPE;
+
+export interface PluginDeps {
+  discover: Pick<DiscoverStart, 'urlGenerator'>;
+}
+
+export interface CoreDeps {
+  application: Pick<CoreStart['application'], 'navigateToApp'>;
+}
+
+export interface Params {
+  start: StartServicesGetter<PluginDeps, unknown, CoreDeps>;
+}
+
+export class ExploreDataContextMenuAction implements Action<EmbeddableContext> {
+  public readonly id = ACTION_EXPLORE_DATA;
+
+  public readonly type = ACTION_EXPLORE_DATA;
+
+  public readonly order = 200;
+
+  constructor(private readonly params: Params) {}
+
+  public getDisplayName() {
+    return i18n.translate('xpack.discover.FlyoutCreateDrilldownAction.displayName', {
+      defaultMessage: 'Explore underlying data',
+    });
+  }
+
+  public getIconType() {
+    return 'discoverApp';
+  }
+
+  public async isCompatible({ embeddable }: EmbeddableContext) {
+    if (!this.params.start().plugins.discover.urlGenerator) return false;
+    if (!isVisualizeEmbeddable(embeddable)) return false;
```

Comment:
We also want to enable this action for TSVB and Lens. Both of them have a shared challenge that they can be backed by more than one index pattern, AFAIK. And that is an open question, what do we do in that case? And then more specifically for each visualization:

For TSVB we cannot do it now, as we need KibanaApp to add support for `ui_actions` in TSVB. I know they are working on it now, and I'll check what else, if anything, needs to be done after they add `ui_actions` support to TSVB.

For Lens I believe we can already enable it. But I would need to check that and replace this line by something that would cover embeddables in general. And of course there is this question of what we want to do when there are multiple index patterns. So, I'm thinking it would be probably cleaner to add support for Lens in a follow up PR.